### ISSUE TITLE
persistent peers dialing management fix

### ISSUE THREAD

**Issue by dlguddus** [tendermint/tendermint#2542]
case study

- one of sentry was dead for more than 20min
- dead sentry became alive after dead period
- a relay has the sentry as a persistent peer
- relay node config : pex/seedmode false

tested result

- relay never retry dialing the sentry after the dead sentry alive
- rpc dial_peer from relay to the sentry throw error : Permanently Removed
- only way to re-establish connection is to restart the relay node, or rpc-dial relay node from sentry

problem

- although sentry had been offline for some time, relay should try reconnect the sentry at least every 1 minute. no harm to do it. 
- also it is fair based on meaning of PERSISTENT. frequency can be configured in config.toml.
- gaiad should never prohibit users to manually dial a peer via rpc endpoint. human does that when he has enough good reason to do it.

opinion

- it is quite a urgent thing to fix because it affects a validator’s and the whole network’s connection stability. especially, prohibiting human manual rpc-dialing is malfunctioned in my opinion.

solution

- dial persistent peers “FOREVER” with given frequency(default every 1 minute)
- don’t prohibit RPC dial_peer in ANY CIRCUMSTANCES

**Comment by dlguddus** [comment]
> Are you running with --rpc.unsafe on?

Yes. the error message was : "Permanently Removed", not 404.
I tried it on localhost. It worked with other peers.

> Is it possible it was in the exponential backoff phase?

Yes I suppose. Because offline period was about 20 minutes so it was definitely already backed off.
I waited more than 10 minutes but I am not sure how long it was.

> but then if the peer is genuinely gone, there should be a way to tell the node to stop dialing it

There is no certain way to decide the peer is genuinely gone. It can suddenly up after 5 hours of offline.
In a matter of persistent peers, it is the node runner's responsibility to actively remove or add persistent peers on their hands. We already do that with automated scripts.

Because of exponential backoff, it is difficult for us to automatically connect to just awake persistent peer as soon as possible. Still, by rpc dialing, we can do "dialing forever" with API script, but it is also prohibited by gaiad. (Permanently Removed error)

**Comment by melekes** [comment]
> Yes. the error message was : "Permanently Removed", not 404.

can you provide some logs? There's no such code as Permanently Removed https://en.wikipedia.org/wiki/List_of_HTTP_status_codes. Maybe "301 Permanently Moved" is what you mean? If yes, then we should just configure our client to follow redirects (`curl -L google.com`)?

> But when the peer is defined as persistent, it should be connected as soon as it becomes online again.

That's hard to get right in one direction. If the peer, who's woke up, send us a message ("Hey, I am back into business again"), then it would be much faster than exp. backoff

**Comment by dlguddus** [comment]
> Maybe "301 Permanently Moved" is what you mean?

I couldn't replay the error at this time. "Permanently Moved" could be right. It might be my wrong memory. Sorry. So I tested that RPC dial is doable without problem now.

> If the peer, who's woke up, send us a message ("Hey, I am back into business again"), then it would be much faster than exp. backoff

Yes it is true, but unfortunately my sentry does not have relay's IP. Only relay has sentry's IP, so it is only one-way dialing available.

Anyway, I found that RPC dialing disability was just from my misunderstanding so there is no need for me to hold this issue further. So I close this issue. Thank you for your replies.
